### PR TITLE
Report: group event listener extended info by call site location

### LINKED ISSUE #959: Event listener audits show duplicate line/col entries
@brendankenny talked through this today. Turns out the duplicates are actually correct. The reason the line/col are duplicated is because the location of where addEventListener is invoked is different than where the user's event handler is defined. This can happen for example, when the user uses a library. The user hands off a callback and the library attached the listener in another place in the source.

screen shot 2016-11-16 at 10 30 46 pm

The fix is to group the constructed event listener code snippets under their common line/col buckets. Meaning show the url/line/col once, and the list of snippets below it.

### PR DESCRIPTION
**Pull request by ebidel**
R: @brendankenny 

The fix for #959 with a documented explanation of it. Tried to comment the code as much as possible :)

### CODE REVIEW

File: lighthouse-core/lib/event-helpers.js [hunk 1]
```diff
@@ -34,6 +34,52 @@ function addFormattedCodeSnippet(listener) {
   }, listener);
 }
 
+/**
+ * Groups event listeners under url/line/col "violation buckets".
+ *
+ * @param {!Array<Object>} listeners Results from the event listener gatherer.
+ * @return {!Array<Object>}
+ *     A copy of the original listener object with the added properties.
```
Comment (brendankenny) [inline lighthouse-core/lib/event-helpers.js]: not really a  copy of the original object anymore

File: lighthouse-core/lib/event-helpers.js [hunk 2]
```diff
@@ -34,6 +34,52 @@ function addFormattedCodeSnippet(listener) {
...
+  for (const item of map.entries()) {
+    const lineColUrlObj = JSON.parse(item[0]);
+    const listenersForLocation = item[1];
+    // Aggregate the code snippets.
+    const codeSnippets = listenersForLocation.reduce((prev, loc) => {
+      return prev + loc.code.trim() + '\n\n';
+    }, '');
+    lineColUrlObj.code = codeSnippets;
+    // All listeners under this bucket have the same line/col. We use the first
+    // one as the label for all of them.
```
Comment (brendankenny) [inline lighthouse-core/lib/event-helpers.js]: maybe `We use the first's label as the label for all of them.`?

File: lighthouse-core/lib/event-helpers.js [hunk 3]
```diff
@@ -34,6 +34,52 @@ function addFormattedCodeSnippet(listener) {
   }, listener);
 }
 
+/**
+ * Groups event listeners under url/line/col "violation buckets".
+ *
+ * @param {!Array<Object>} listeners Results from the event listener gatherer.
+ * @return {!Array<Object>}
```
Comment (brendankenny) [inline lighthouse-core/lib/event-helpers.js]: `@return {!Array<{line: number, col: number, url: string, type: string, code: string, label: string}>}`? :)

File: lighthouse-core/audits/dobetterweb/no-mutation-events.js [hunk 4]
```diff
@@ -75,12 +75,14 @@ class NoMutationEventsAudit extends Audit {
...
     const pageHost = url.parse(artifacts.URL.finalUrl).host;
 
-    const results = listeners.filter(loc => {
+    let results = listeners.filter(loc => {
       const isMutationEvent = this.MUTATION_EVENTS.indexOf(loc.type) !== -1;
       const sameHost = loc.url ? url.parse(loc.url).host === pageHost : true;
       return sameHost && isMutationEvent;
     }).map(EventHelpers.addFormattedCodeSnippet);
 
+    results = EventHelpers.groupCodeSnippetsByLocation(results);
```
Comment (ebidel) [inline lighthouse-core/audits/dobetterweb/no-mutation-events.js]: done

File: lighthouse-core/audits/dobetterweb/uses-passive-event-listeners.js [hunk 5]
```diff
@@ -76,6 +76,8 @@ class PassiveEventsAudit extends Audit {
              !mentionsPreventDefault;
     }).map(EventHelpers.addFormattedCodeSnippet);
 
+    results = EventHelpers.groupCodeSnippetsByLocation(results);
```
Comment (ebidel) [inline lighthouse-core/audits/dobetterweb/uses-passive-event-listeners.js]: done

File: lighthouse-core/lib/event-helpers.js (hunk 3 above)
Comment (ebidel) [inline lighthouse-core/lib/event-helpers.js]: done

File: lighthouse-core/lib/event-helpers.js [hunk 6]
```diff
@@ -34,6 +34,52 @@ function addFormattedCodeSnippet(listener) {
   }, listener);
 }
 
+/**
+ * Groups event listeners under url/line/col "violation buckets".
+ *
+ * @param {!Array<Object>} listeners Results from the event listener gatherer.
```
Comment (ebidel) [inline lighthouse-core/lib/event-helpers.js]: done

File: lighthouse-core/lib/event-helpers.js (hunk 1 above)
Comment (ebidel) [inline lighthouse-core/lib/event-helpers.js]: done

File: lighthouse-core/lib/event-helpers.js [hunk 7]
```diff
@@ -34,6 +34,52 @@ function addFormattedCodeSnippet(listener) {
...
+ * @return {!Array<Object>}
+ *     A copy of the original listener object with the added properties.
+ */
+function groupCodeSnippetsByLocation(listeners) {
+  // The listener gatherer returns a list of (url/line/col) src locations where
```
Comment (ebidel) [inline lighthouse-core/lib/event-helpers.js]: done

File: lighthouse-core/lib/event-helpers.js [hunk 8]
```diff
@@ -34,6 +34,52 @@ function addFormattedCodeSnippet(listener) {
...
+  // event handlers were attached to DOM nodes. This location is where
+  // addEventListener was invoked, but it's not guaranteed to be where
+  // the user's event handler was defined. An example is libraries, where the
+  // user provides a callback and the library calls addEventListener (another
+  // part of the codebase). Instead of showing a repetitive list of
+  // url/line/col violations, we make each trio a unique bucket and
```
Comment (ebidel) [inline lighthouse-core/lib/event-helpers.js]: done

File: lighthouse-core/lib/event-helpers.js [hunk 9]
```diff
@@ -34,6 +34,52 @@ function addFormattedCodeSnippet(listener) {
...
+  // display the user's event handlers under each.
+  const map = new Map();
```
Comment (ebidel) [inline lighthouse-core/lib/event-helpers.js]: done

File: lighthouse-core/lib/event-helpers.js [hunk 10]
```diff
@@ -34,6 +34,52 @@ function addFormattedCodeSnippet(listener) {
...
+  listeners.forEach(loc => {
+    const key = JSON.stringify(
+        {line: loc.line, col: loc.col, url: loc.url, type: loc.type});
```
Comment (ebidel) [inline lighthouse-core/lib/event-helpers.js]: yea < 100. done

File: lighthouse-core/lib/event-helpers.js [hunk 11]
```diff
@@ -34,6 +34,52 @@ function addFormattedCodeSnippet(listener) {
...
+      map.get(key).push(loc);
+    } else {
+      map.set(key, [loc]);
+    }
+  });
+
+  const results = [];
+  for (const item of map.entries()) {
+    const lineColUrlObj = JSON.parse(item[0]);
+    const listenersForLocation = item[1];
```
Comment (ebidel) [inline lighthouse-core/lib/event-helpers.js]: done

File: lighthouse-core/lib/event-helpers.js (hunk 2 above)
Comment (ebidel) [inline lighthouse-core/lib/event-helpers.js]: done

**Comment by ebidel** [comment]
PTAL

**Pull request merged**